### PR TITLE
R-package RNN refactor

### PR DESCRIPTION
**Pull request by jeremiedb**
New to pull with corrected submodules. 

Related to #5488, this is to provide a more flexible and efficient framework to run RNN within the R-package. 

! This pull does brake the current RNN API - better to renamed files/function to maintain legacy support? 
 
- Built around `mx.symbol.RNN`
- Support bucketing and masking through: `mx.io.bucket.iter.R`
- Support multi-device training, single instance inference. 
- Only support seq-to-one at the moment but inference on one-to-one should be added shortly
- Requires to run with CUDA until `mx.symbol.RNN` gets a CPU support

End to end tutorial can be seen [here](https://github.com/jeremiedb/mxnet_R_bucketing). Will be finalized in a blog post when one-to-one inference suport is completed. 

@thirdwing

### CODE REVIEW

File: R-package/R/mx.io.bucket.iter.R [hunk 1]
```diff
@@ -0,0 +1,110 @@
+library(mxnet)
```
Comment (thirdwing) [inline R-package/R/mx.io.bucket.iter.R]: Please remove "library(mxnet)".

File: R-package/R/rnn.R [hunk 2]
```diff
@@ -1,342 +1,101 @@
...
-                                    bias=param$i2h.bias,
-                                    num.hidden=num.hidden,
-                                    name=paste0("t", seqidx, ".l", layeridx, ".i2h"))
-    h2h <- mx.symbol.FullyConnected(data=prev.state$h,
-                                    weight=param$h2h.weight,
-                                    bias=param$h2h.bias,
-                                    num.hidden=num.hidden,
-                                    name=paste0("t", seqidx, ".l", layeridx, ".h2h"))
-    hidden <- i2h + h2h
+library(mxnet)
```
Comment (thirdwing) [inline R-package/R/rnn.R]: Please remove "library(mxnet)".

File: R-package/R/rnn.graph.R [hunk 3]
```diff
@@ -0,0 +1,123 @@
+library(mxnet)
```
Comment (thirdwing) [inline R-package/R/rnn.graph.R]: Please remove "library(mxnet)".

File: R-package/R/rnn.infer.R [hunk 4]
```diff
@@ -0,0 +1,77 @@
+library(mxnet)
+
+source("rnn.R")
```
Comment (thirdwing) [inline R-package/R/rnn.infer.R]: Please remove these two lines.

File: R-package/R/rnn.graph.R [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+  if (output_last_state){
+    # group <- mx.symbol.Group(c(unlist(last.states), loss))
+    # return(group)
+    return(loss)
+  } else return(loss)
+}
+
+
+
+# data <- mx.symbol.Variable("data")
```
Comment (thirdwing) [inline R-package/R/rnn.graph.R]: Can we remove the testing code?

**Comment by jeremiedb** [comment]
@thirdwing `source()` and `library()` calls removed. 

Functions `mx.model.train.rnn.buckets` and `mx.rnn.buckets` merged into `model.rnn.R` in order to better align with `model.R`.

**Comment by jeremiedb** [comment]
rnn.model.R is compatible with regular iterator and non RNN models. 
However, the need to bind at each batch results in a slower training time of about 15-20%. Also, aux.params doesn't behave as expected, so shouldn't use BatchNorm in a model trained with mx.rnn.buckets - yet to figure how to correct this.  

Example of application with [one-to-one](https://github.com/jeremiedb/mxnet_R_bucketing/blob/master/README_one_to_one.md) and [seq-to-one](https://github.com/jeremiedb/mxnet_R_bucketing/blob/master/README.md) updated.  

Several backend functionnalities were brought up explicitly in the  mx.rnn.buckets. I think it makes it more transparent to track treatment of the arg.params.

**Comment by piiswrong** [comment]
@thirdwing What's the status?

File: R-package/R/mx.io.bucket.iter.R [hunk 6]
```diff
@@ -0,0 +1,110 @@
+
+BucketIter <- setRefClass("BucketIter", fields = c("buckets", "bucket.names", "batch.size", 
+                                                   "data.mask.element", "shuffle", "bucket.plan", "bucketID", "epoch", "batch", "batch.per.bucket", 
+                                                   "last.batch.pad", "batch.per.epoch", "seed"), contains = "Rcpp_MXArrayDataIter",
```
Comment (thirdwing) [inline R-package/R/mx.io.bucket.iter.R]: Removing `contains = "Rcpp_MXArrayDataIter",` should make CI passed.

**Comment by thirdwing** [comment]
Sorry, a little busy after back to school. I will review this in the weekend.

**Comment by jeremiedb** [comment]
Improve the harmonization with model.FeedForward: same fixed.params, arg.params and aux.params input arguments. 

Remove redundancies between model.buckets and model.train.buckets. 

model.buckets is as performant on single symbol model as model.Feedforward. If using BatchNorm, single symbol works fine but the still is issue if training on list of symbols. 

Comment out the lstm in testthat to pass CI.

Integrate CPU compatible RNN construction with raw lstm and gru cells in rnn.graph. No support for masking. Still need to test efficiency of inference and potentially adapt the inference functions. 

Unsure if wouldn't be preferable to assume a batch.size X seq.length input dimension to interator as it's the format expectged by symnol.RNN cell. Or add a shape detector to handle it automatically as in the python API.

**Comment by thirdwing** [comment]
@jeremiedb Great work! I think this can be merged after some very minor tweaks:

(1) you don' need any `mxnet:::`, since your code is part of this package;

(2) let's remove `test_lstm.R` for now. We can add new test for RNN/LSTM later.

**Comment by jeremiedb** [comment]
@thirdwing I'm out for 10 days, I'll try to put my hands a computer by then but it may take a few days.

**Comment by thirdwing** [comment]
The CI test failed with one python testing. Not related to this PR.

**Comment by thirdwing** [comment]
@jeremiedb Looking forward to the blog post.

**Comment by thirdwing** [comment]
@jeremiedb Can you update with master branch? Let's see the CI test.

**Comment by jeremiedb** [comment]
@thirdwing CI has issue with some Python test, not related to this pull.
